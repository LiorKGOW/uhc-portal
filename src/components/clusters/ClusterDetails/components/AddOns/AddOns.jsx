import React from 'react';
import PropTypes from 'prop-types';

import {
  Card,
  CardHead,
  CardActions,
  CardBody,
  Button,
  EmptyState,
  EmptyStateBody,
  Title,
  EmptyStateIcon,
} from '@patternfly/react-core';
import { Spinner } from '@redhat-cloud-services/frontend-components';
import { IntegrationIcon } from '@patternfly/react-icons';

import clusterStates from '../../../common/clusterStates';
import ErrorBox from '../../../../common/ErrorBox';

import {
  getInstallState,
  isInstalled,
  hasQuota,
  availableAddOns,
} from './AddOnsHelper';

class AddOns extends React.Component {
  componentDidMount() {
    const {
      clusterID,
      getClusterAddOns,
      clusterAddOns,
    } = this.props;
    if (clusterAddOns.clusterID !== clusterID || (!clusterAddOns.pending)) {
      getClusterAddOns(clusterID);
    }
  }

  componentDidUpdate(prevProps) {
    const {
      clusterID,
      getClusterAddOns,
      clusterAddOns,
      addClusterAddOnResponse,
      getOrganizationAndQuota,
    } = this.props;
    if ((addClusterAddOnResponse.fulfilled && prevProps.addClusterAddOnResponse.pending)
        && !clusterAddOns.pending) {
      // Fetch cluster add-ons again if we just added a cluster add-on
      getClusterAddOns(clusterID);
      // Refresh quota after installing add-ons
      getOrganizationAndQuota();
    }
  }

  componentWillUnmount() {
    const { clearClusterAddOnsResponses } = this.props;
    clearClusterAddOnsResponses();
  }

  render() {
    const {
      addOns,
      cluster,
      clusterAddOns,
      clusterID,
      addClusterAddOn,
      addClusterAddOnResponse,
      organization,
      quota,
    } = this.props;

    if (clusterAddOns.pending || addClusterAddOnResponse.pending) {
      return (
        <EmptyState>
          <EmptyStateBody>
            <Spinner centered />
          </EmptyStateBody>
        </EmptyState>
      );
    }

    // Add-ons can only be installed when the cluster is ready
    const isClusterReady = cluster.state === clusterStates.READY;
    const availableAddOnsList = availableAddOns(addOns, clusterAddOns, organization, quota);
    const hasAddOns = availableAddOnsList.length > 0;

    if (!hasAddOns) {
      return (
        <EmptyState>
          <EmptyStateIcon icon={IntegrationIcon} />
          {addOns.error && (
            <ErrorBox message="Error getting add-ons" response={addOns} />
          )}
          <Title headingLevel="h5" size="lg">No add-ons available for this cluster</Title>
          <EmptyStateBody>
            There are no add-ons available for this cluster.
          </EmptyStateBody>
        </EmptyState>
      );
    }

    if (clusterAddOns.error) {
      return (
        <EmptyState>
          <ErrorBox message="Error getting cluster add-ons" response={clusterAddOns} />
        </EmptyState>
      );
    }

    return (
      <div className="cluster-details-addons-tab-contents">
        { addClusterAddOnResponse.error && (
        <ErrorBox message="Error adding add-ons" response={addClusterAddOnResponse} />
        )}
        { availableAddOnsList.map(addOn => (
          <Card key={addOn.id}>
            <CardHead>
              <Title headingLevel="h2" size="2xl">{addOn.name}</Title>
              <CardActions>
                { (isInstalled(addOn, clusterAddOns) && getInstallState(addOn, clusterAddOns)) || (
                <Button
                  variant="secondary"
                  aria-label="Install"
                  isDisabled={
                    addClusterAddOnResponse.pending
                      || !isClusterReady
                      || !cluster.canEdit
                      || !hasQuota(addOn, organization, quota)
                  }
                  onClick={() => addClusterAddOn(clusterID, addOn.id)}
                >
                  Install
                </Button>
                )}
              </CardActions>
            </CardHead>
            <CardBody className="addon-card">
              <div className="addon-icon">
                { addOn.icon && (
                <img alt={addOn.name} src={`data:image/png;base64,${addOn.icon}`} />
                )}
              </div>
              <div className="addon-description">
                {addOn.description}
              </div>
            </CardBody>
          </Card>
        ))}
      </div>
    );
  }
}

AddOns.propTypes = {
  clusterID: PropTypes.string.isRequired,
  cluster: PropTypes.object.isRequired,
  addOns: PropTypes.object.isRequired,
  clusterAddOns: PropTypes.object.isRequired,
  organization: PropTypes.object.isRequired,
  quota: PropTypes.object.isRequired,
  getOrganizationAndQuota: PropTypes.func.isRequired,
  getClusterAddOns: PropTypes.func.isRequired,
  addClusterAddOn: PropTypes.func.isRequired,
  addClusterAddOnResponse: PropTypes.object.isRequired,
  clearClusterAddOnsResponses: PropTypes.func.isRequired,
};

export default AddOns;
