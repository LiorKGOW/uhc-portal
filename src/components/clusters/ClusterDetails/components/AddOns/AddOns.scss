.addon-delete-modal-textinput {
  font-weight: var(--pf-global--FontWeight--semi-bold);
}

.addon-parameter-default-button {
  padding-left: 0
}