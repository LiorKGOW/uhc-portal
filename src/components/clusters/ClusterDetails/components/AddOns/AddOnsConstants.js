const GET_ADDONS = 'GET_ADDONS';
const GET_CLUSTER_ADDONS = 'GET_CLUSTER_ADDONS';
const ADD_CLUSTER_ADDON = 'ADD_CLUSTER_ADDON';
const UPDATE_CLUSTER_ADDON = 'UPDATE_CLUSTER_ADDON';
const DELETE_CLUSTER_ADDON = 'DELETE_CLUSTER_ADDON';
const CLEAR_CLUSTER_ADDON_RESPONSES = 'CLEAR_CLUSTER_ADDON_RESPONSES';

const AddOnsConstants = {
  GET_ADDONS,
  GET_CLUSTER_ADDONS,
  UPDATE_CLUSTER_ADDON,
  ADD_CLUSTER_ADDON,
  DELETE_CLUSTER_ADDON,
  CLEAR_CLUSTER_ADDON_RESPONSES,
  INSTALLATION_STATE: {
    PENDING: 'pending',
    INSTALLING: 'installing',
    UPDATING: 'updating',
    DELETING: 'deleting',
    DELETED: 'deleted',
    FAILED: 'failed',
    READY: 'ready',
  },
};

export default AddOnsConstants;
