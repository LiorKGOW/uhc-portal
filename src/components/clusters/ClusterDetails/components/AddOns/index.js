import { connect } from 'react-redux';
import AddOns from './AddOns';
import {
  getAddOns,
  getClusterAddOns,
  addClusterAddOn,
  clearClusterAddOnsResponses,
} from './AddOnsActions';
import { getOrganizationAndQuota } from '../../../../../redux/actions/userActions';
import { modalActions } from '../../../../common/Modal/ModalActions';
import { getMachinePools } from '../MachinePools/MachinePoolsActions';

const mapStateToProps = (state) => ({
  addOns: state.addOns.addOns,
  cluster: state.clusters.details.cluster,
  clusterAddOns: state.addOns.clusterAddOns,
  clusterMachinePools: state.machinePools.getMachinePools,
  addClusterAddOnResponse: state.addOns.addClusterAddOnResponse,
  updateClusterAddOnResponse: state.addOns.updateClusterAddOnResponse,
  deleteClusterAddOnResponse: state.addOns.deleteClusterAddOnResponse,
  organization: state.userProfile.organization,
  quota: state.userProfile.organization.quotaList,
});

const mapDispatchToProps = {
  getOrganizationAndQuota,
  getAddOns,
  getClusterAddOns,
  getMachinePools,
  addClusterAddOn,
  clearClusterAddOnsResponses,
  openModal: modalActions.openModal,
  closeModal: modalActions.closeModal,
};

export default connect(mapStateToProps, mapDispatchToProps)(AddOns);
