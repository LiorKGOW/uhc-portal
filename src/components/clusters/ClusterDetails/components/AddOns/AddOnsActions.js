import get from 'lodash/get';
import AddOnsConstants from './AddOnsConstants';
import { clusterService } from '../../../../../services';

const getAddOns = (clusterID) => (dispatch) =>
  dispatch({
    type: AddOnsConstants.GET_ADDONS,
    payload: clusterService.getAddOns(clusterID).then((response) => {
      const items = get(response, 'data.items', []);
      return {
        items,
        resourceNames: items.map((addOn) => addOn.resource_name),
      };
    }),
  });

const getClusterAddOns = (clusterID) => (dispatch) =>
  dispatch({
    type: AddOnsConstants.GET_CLUSTER_ADDONS,
    payload: clusterService.getClusterAddOns(clusterID).then((response) => {
      const items = get(response, 'data.items', []);
      return {
        clusterID,
        items,
      };
    }),
  });

const addClusterAddOn = (clusterID, addOnData) => (dispatch) =>
  dispatch({
    type: AddOnsConstants.ADD_CLUSTER_ADDON,
    payload: clusterService.addClusterAddOn(clusterID, addOnData),
  });

const updateClusterAddOn = (clusterID, addOnID, addOnData) => (dispatch) =>
  dispatch({
    type: AddOnsConstants.UPDATE_CLUSTER_ADDON,
    payload: clusterService.updateClusterAddOn(clusterID, addOnID, addOnData).then((response) => {
      dispatch(getClusterAddOns(clusterID));
      return response;
    }),
  });

const deleteClusterAddOn = (clusterID, addOnData) => (dispatch) =>
  dispatch({
    type: AddOnsConstants.DELETE_CLUSTER_ADDON,
    payload: clusterService.deleteClusterAddOn(clusterID, addOnData),
  });

const clearClusterAddOnsResponses = () => ({
  type: AddOnsConstants.CLEAR_CLUSTER_ADDON_RESPONSES,
});

const addOnsActions = {
  getAddOns,
  getClusterAddOns,
  addClusterAddOn,
  updateClusterAddOn,
  deleteClusterAddOn,
  clearClusterAddOnsResponses,
};

export {
  addOnsActions,
  getAddOns,
  getClusterAddOns,
  addClusterAddOn,
  updateClusterAddOn,
  deleteClusterAddOn,
  clearClusterAddOnsResponses,
};
