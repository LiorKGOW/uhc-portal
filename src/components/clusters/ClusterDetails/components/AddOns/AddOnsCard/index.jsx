import { connect } from 'react-redux';
import AddOnsCard from './AddOnsCard';
import { addClusterAddOn } from '../AddOnsActions';
import { modalActions } from '../../../../../common/Modal/ModalActions';

const mapStateToProps = state => ({
  cluster: state.clusters.details.cluster,
  addClusterAddOnResponse: state.addOns.addClusterAddOnResponse,
});

const mapDispatchToProps = {
  addClusterAddOn,
  openModal: modalActions.openModal,
  closeModal: modalActions.closeModal,
};

export default connect(mapStateToProps, mapDispatchToProps)(AddOnsCard);
