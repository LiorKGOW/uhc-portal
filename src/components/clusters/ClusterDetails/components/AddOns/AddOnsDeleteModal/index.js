import { connect } from 'react-redux';
import shouldShowModal from '../../../../../common/Modal/ModalSelectors';
import { closeModal } from '../../../../../common/Modal/ModalActions';
import AddOnsDeleteModal from './AddOnsDeleteModal';
import { addOnsActions } from '../AddOnsActions';


const mapStateToProps = state => ({
  isOpen: shouldShowModal(state, 'add-ons-delete-modal'),
  modalData: state.modal.data,
  deleteClusterAddOnResponse: state.addOns.deleteClusterAddOnResponse,
});

const mapDispatchToProps = {
  deleteClusterAddOn: (clusterID, addOnID) => addOnsActions.deleteClusterAddOn(clusterID, addOnID),
  closeModal: () => closeModal(),
  clearDeleteAddOnResponse: () => addOnsActions.clearClusterAddOnsResponses(),
};

export default connect(mapStateToProps, mapDispatchToProps)(AddOnsDeleteModal);
