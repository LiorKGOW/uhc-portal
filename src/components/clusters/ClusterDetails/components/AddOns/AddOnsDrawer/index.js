import { connect } from 'react-redux';
import AddOnsDrawer from './AddOnsDrawer';
import { addClusterAddOn } from '../AddOnsActions';
import { modalActions } from '../../../../../common/Modal/ModalActions';

const mapStateToProps = state => ({
  cluster: state.clusters.details.cluster,
  addClusterAddOnResponse: state.addOns.addClusterAddOnResponse,
  deleteClusterAddOnResponse: state.addOns.deleteClusterAddOnResponse,
  submitClusterAddOnResponse: state.modal.data.isUpdateForm
    ? state.addOns.updateClusterAddOnResponse
    : state.addOns.addClusterAddOnResponse,
});

const mapDispatchToProps = {
  addClusterAddOn,
  openModal: modalActions.openModal,
};

export default connect(mapStateToProps, mapDispatchToProps)(AddOnsDrawer);
