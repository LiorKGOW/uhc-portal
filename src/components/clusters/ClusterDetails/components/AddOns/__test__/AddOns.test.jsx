import React from 'react';
import { shallow } from 'enzyme';

import AddOns from '../AddOns';
import { mockAddOns, mockClusterAddOns } from './AddOns.fixtures';
import { addonsQuota } from '../../../../common/__test__/quota.fixtures';
import fixtures from '../../../__test__/ClusterDetails.fixtures';

describe('<AddOns />', () => {
  let wrapper;
  const getClusterAddOns = jest.fn();
  const clearClusterAddOnsResponses = jest.fn();
  const addClusterAddOnResponse = {};
  const updateClusterAddOnResponse = {};
  const deleteClusterAddOnResponse = {};
  const getOrganizationAndQuota = jest.fn();
  const openModal = jest.fn();

  const { clusterDetails, organization } = fixtures;

  beforeAll(() => {
    wrapper = shallow(<AddOns
      clusterID={clusterDetails.cluster.id}
      cluster={clusterDetails.cluster}
      organization={organization}
      addOns={mockAddOns}
      clusterAddOns={mockClusterAddOns}
      clusterMachinePools={{}}
      quota={addonsQuota}
      getOrganizationAndQuota={getOrganizationAndQuota}
      getClusterAddOns={getClusterAddOns}
      addClusterAddOnResponse={addClusterAddOnResponse}
      updateClusterAddOnResponse={updateClusterAddOnResponse}
      deleteClusterAddOnResponse={deleteClusterAddOnResponse}
      clearClusterAddOnsResponses={clearClusterAddOnsResponses}
      openModal={openModal}
    />);
  });

  it('should render', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
