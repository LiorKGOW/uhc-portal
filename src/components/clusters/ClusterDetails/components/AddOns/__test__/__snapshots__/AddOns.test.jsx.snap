// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddOns /> should render 1`] = `
<Fragment>
  <Gallery
    className="addon-gallery"
    hasGutter={true}
  >
    <GalleryItem>
      <Connect(AddOnsCard)
        addOn={
          Object {
            "description": "Based on the open source Istio project, Red Hat OpenShift Service Mesh provides an easy way to create a network of deployed services that provides discovery, load balancing, service-to-service authentication, failure recovery, metrics, and monitoring.",
            "enabled": true,
            "href": "/api/clusters_mgmt/v1/addons/service-mesh",
            "icon": null,
            "id": "service-mesh",
            "install_mode": "single_namespace",
            "kind": "AddOn",
            "label": "api.openshift.com/addon-service-mesh",
            "name": "Red Hat OpenShift Service Mesh",
            "operator_name": "service-mesh-operator",
            "resource_cost": 1,
            "resource_name": "addon-service-mesh",
            "target_namespace": "redhat-service-mesh-operator",
          }
        }
        hasQuota={false}
        installedAddOn={
          Object {
            "addon": Object {
              "id": "service-mesh",
            },
            "href": "/api/clusters_mgmt/v1/addons/service-mesh",
            "kind": "AddOnLink",
            "state": "ready",
          }
        }
        key="service-mesh"
        requirements={
          Object {
            "errorMsgs": Array [],
            "fulfilled": true,
          }
        }
      />
    </GalleryItem>
  </Gallery>
  <Connect(ReduxForm)
    clusterID="1i4counta3holamvo1g5tp6n8p3a03bq"
  />
  <Connect(AddOnsDeleteModal) />
</Fragment>
`;
